types/decimal.go

```diff
@@ -0,0 +1,356 @@
+package types
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"strconv"
+	"strings"
+	"testing"
+)
+
+// NOTE: never use new(Dec) or else we will panic unmarshalling into the
+// nil embedded big.Int
+type Dec struct {
+	*big.Int `json:"int"`
+}
+
+// number of decimal places
+const Precision = 10
+
+var (
+	precisionExpReuse    = big.NewInt(Precision)
+	ten                  = big.NewInt(10)
+	precisionReuse       = new(big.Int).Exp(big.NewInt(10), big.NewInt(Precision), nil)
+	precisionMultipliers []*big.Int
+)
+
+// Set precision multipliers
+func init() {
+	precisionMultipliers = make([]*big.Int, Precision+1)
+	for i := 0; i <= Precision; i++ {
+		precisionMultipliers[i] = calcPrecisionMultiplier(int64(i))
+	}
+}
+
+func precisionInt() *big.Int {
+	return new(big.Int).Set(precisionReuse)
+}
+
+// nolint - common values
+func ZeroDec() Dec { return Dec{big.NewInt(0)} }
+func OneDec() Dec  { return Dec{precisionInt()} }
+
+// calculate the precision multiplier
+func calcPrecisionMultiplier(prec int64) *big.Int {
+	if prec > Precision {
+		panic("too much precision")
+	}
+	zerosToAdd := Precision - prec
+	multiplier := new(big.Int).Exp(big.NewInt(10), big.NewInt(zerosToAdd), nil)
+	return multiplier
+}
+
+// get the precision multiplier. Do not mutate result.
+func precisionMultiplier(prec int64) *big.Int {
+	if prec > Precision {
+		panic("too much precision")
+	}
+	return precisionMultipliers[prec]
+}
+
+// create a new Dec from integer assuming whole numbers
+// CONTRACT: prec <= Precision
+func NewDec(i, prec int64) Dec {
+	return Dec{
+		new(big.Int).Mul(big.NewInt(i), precisionMultiplier(prec)),
+	}
+}
+
+// create a new Dec from big integer assuming whole numbers
+// CONTRACT: prec !> Precision
+func NewDecFromBigInt(i *big.Int, prec int64) Dec {
+	return Dec{
+		new(big.Int).Mul(i, precisionMultiplier(prec)),
+	}
+}
+
+// create a new Dec from big integer assuming whole numbers
+// CONTRACT: prec !> Precision
+func NewDecFromInt(i Int, prec int64) Dec {
+	return Dec{
+		new(big.Int).Mul(i.BigInt(), precisionMultiplier(prec)),
+	}
+}
+
+// create a decimal from a decimal string (ex. "1234.5678")
+func NewDecFromStr(str string) (d Dec, err Error) {
```

Comment:
This function needs to be fuzzed (since it sanitizes user input)

Reply:
Fuzzing this isn't critical, since we validate all returned values makes sense.

Reply:
I second val